### PR TITLE
Channel logic state separation

### PR DESCRIPTION
**Pull request by leandroBorgesFerreira**
issue: https://github.com/GetStream/stream-chat-android/issues/3730

### 🎯 Goal

Separate ChannelLogic into database and state logic, so it is possible to move classes related to sending attachments to LLC module instead of OfflinePlugin module and actually moving those classes.

### 🛠 Implementation details

Although this PR is very big, most of it are licenses, unit tests (a large part), packaging move, some refactor so static analyses of code pass without changing the baselin, imports change due to classes being moved and the move of classes. The real part where code is really changed is inside `ChannelLogic` and `ChannelStateLogic`. Changes were: 

- Taking state away from `ChannelLogic`, now it can consult state of some cases, but it can't write it. 
- Creating `ChannelStateLogicImpl` so it handles the state of a channel. 
- Creating the interface `ChannelStateLogic`. This interface is available in `ChannelStateLogic` and the implementation can be provided or not. If the implementation is not provided, then the state is not updated, but the SDK behaves normally. 

### 🧪 Testing

Test channel state in a broad way. 

#### General
- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [ ] Assigned a person / code owner group (required)
- [ ] Thread with the PR link started in a respective Slack channel (#android-chat-core or #android-chat-ui) (required)
- [x] PR targets the `develop` branch
- [x] PR is linked to the GitHub issue it resolves

#### Code & documentation
- ~[ ] Changelog is updated with client-facing changes~. There's no client-facing changes.
- [x] New code is covered by unit tests
- ~[ ] Comparison screenshots added for visual changes~
- ~[ ] Affected documentation updated (KDocs, docusaurus, tutorial)~

### ☑️Reviewer Checklist
- [ ] UI Components sample runs & works
- [ ] Compose sample runs & works
- [ ] UI Changes correct (before & after images)
- [ ] Bugs validated (bugfixes)
- [ ] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

### 🎉 GIF
giphy (1)
_Right? Right? No?? yeah, sorry..._

### CODE REVIEW

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/models/ChannelData.kt
Comment (kanat) [inline stream-chat-android-client/src/main/java/io/getstream/chat/android/client/models/ChannelData.kt:69]: Should we mark it with `@InternalStreamChatApi`?

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/models/ChannelData.kt
Comment (kanat) [inline stream-chat-android-client/src/main/java/io/getstream/chat/android/client/models/ChannelData.kt:99]: same here

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/channel/state/ChannelMutableState.kt
Comment (leandroBorgesFerreira) [reply stream-chat-android-client/src/main/java/io/getstream/chat/android/client/channel/state/ChannelMutableState.kt]: We can keep this as an internal interface and move and change visibility later, as we discussed. With an internal interface keeping the original names can be good to avoid increasing `ChannelMutableStateImpl` now. What do you think?

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/message/attachments/internal/UploadAttachmentsAndroidWorker.kt
Comment (adasiewiczr) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/message/attachments/internal/UploadAttachmentsAndroidWorker.kt:60]: do we need to do this companion object and methods public? 🤔

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/message/attachments/internal/UploadAttachmentsWorker.kt
Comment (adasiewiczr) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/message/attachments/internal/UploadAttachmentsWorker.kt:129]: Can we just import `ChannelMutableState`? 🤔

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/plugin/state/ChannelMutableState.kt
Comment (adasiewiczr) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/plugin/state/ChannelMutableState.kt:17]: We should put this class into `*.internal` package

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/plugin/state/channel/ChannelState.kt
Comment (adasiewiczr) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/plugin/state/channel/ChannelState.kt:46]: Why do we need to expose `sortedMessages` here? 🤔

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/plugin/state/ChannelMutableState.kt
Comment (kanat) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/plugin/state/ChannelMutableState.kt:57]: Hey @leandroBorgesFerreira 👋

Could we avoid exposing `MutableStateFlow` and probably replace it with setters as we've done in `MutableGlobalState`?

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/channel/state/ChannelMutableState.kt
Comment (kanat) [reply stream-chat-android-client/src/main/java/io/getstream/chat/android/client/channel/state/ChannelMutableState.kt]: We've already done such a thing in `MutableGlobalState`.
Those properties are meant to be private. Even if those are for internal SDK usage I would avoid providing full control to other parts of the SDK by exposing a `MutableStateFlow`.

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/message/attachments/internal/UploadAttachmentsWorker.kt
Comment (leandroBorgesFerreira) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/message/attachments/internal/UploadAttachmentsWorker.kt]: 🤔  That's the original behavior in the `develop` branch. I didn't try to change the class, it is just a code refactor to make conform with static code analyses. But, now that you say it, an error makes way more sense. Let's change it to an error.

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/plugin/listener/internal/TypingEventListenerImpl.kt
Comment (kanat) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/plugin/listener/internal/TypingEventListenerImpl.kt:99]: Can we use an interface here?

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/plugin/logic/internal/LogicRegistry.kt
Comment (kanat) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/plugin/logic/internal/LogicRegistry.kt:93]: Could we rename it to `channelStateLogic`?

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/channel/state/ChannelMutableState.kt
Comment (leandroBorgesFerreira) [reply stream-chat-android-client/src/main/java/io/getstream/chat/android/client/channel/state/ChannelMutableState.kt]: I took a look at the code and this would be no trivial task because of the implementation details of `ChannelMutableStateImpl`. One example would be the removal of messages: 
```
override fun removeLocalMessage(message: Message) {
        mutableState.setMessages(mutableState.messages.value.associateBy { it.id } - message.id)
    }
```

Every time it would be an association, which would impact the performance, because ChannelState exposes messages as a `List<Message>`, but the implementation uses a `Map<String, Message>`. Let's create a ticket just for that, it can be done right after this PR.

Comment (leandroBorgesFerreira) [reply stream-chat-android-client/src/main/java/io/getstream/chat/android/client/channel/state/ChannelMutableState.kt]: 🤔  Although I can change the other fields, and leave the ones with this map/list relation

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/message/attachments/internal/UploadAttachmentsWorker.kt
Comment (adasiewiczr) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/message/attachments/internal/UploadAttachmentsWorker.kt:35]: can we document it somehow? For example KDoc

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/plugin/state/channel/ChannelState.kt
Comment (adasiewiczr) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/plugin/state/channel/ChannelState.kt]: yeah exactly. We should minimize public API changes during this refactor

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/message/attachments/internal/UploadAttachmentsWorker.kt
Comment (kanat) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/message/attachments/internal/UploadAttachmentsWorker.kt:35]: Let's probably postpone it to the time when gonna work on it?
For now, it's not clearly visible from the code perspective why it's nullable.

**Review by adasiewiczr — APPROVED** [review]
LGTM. Let's wait for second review from @kanat

**Review by kanat — APPROVED** [review]
LGTM

**Comment by kanat** [comment]
I've resolved merge conflicts, however, I would ask @adasiewiczr to look at this PR one more time to make sure the changes as per #3806 are still here and valid.

**Pull request merged**
